mindsdb/integrations/utilities/rag/s3_knowledge_base.py

```diff
@@ -0,0 +1,123 @@
+import re
+from typing import List, Dict,Optional
+
+import tempfile
+import boto3
+from botocore.exceptions import NoCredentialsError
+from mindsdb.interfaces.knowledge_base.data_source_config import S3Config
+
+
+def connect_s3(s3_config: S3Config):
+    """
+    Establishes a connection to AWS S3 using the provided configuration.
+
+    :param s3_config: S3Config object containing AWS credentials and configuration.
+    :return: boto3 S3 client.
+    """
+    try:
+        client_params = {
+            'service_name': 's3',
+            'aws_access_key_id': s3_config.aws_access_key_id,
+            'aws_secret_access_key': s3_config.aws_secret_access_key,
+            'region_name': s3_config.region_name,
+        }
+
+        # Add session token if provided
+        if s3_config.aws_session_token:
+            client_params['aws_session_token'] = s3_config.aws_session_token
+
+        return boto3.client(**client_params)
+    except NoCredentialsError:
+        raise Exception("Credentials not available")
+
+def get_filtered_files(s3_config: S3Config) -> List[Dict]:
+    """
+    Retrieves and filters files from S3 buckets based on the provided configuration.
+
+    :param s3_config: S3Config object containing bucket and file regex patterns.
+    :return: List of dictionaries with filtered file details.
+    """
+    s3 = connect_s3(s3_config)
+    results = []
+
+    # List and filter buckets based on regex patterns
+    all_buckets = [bucket['Name'] for bucket in s3.list_buckets().get('Buckets', [])]
+    filtered_buckets = [bucket for bucket in all_buckets if any(re.search(pattern, bucket) for pattern in s3_config.buckets)]
+
+    for bucket in filtered_buckets:
+        continuation_token = None
+
+        while True:
+            # Fetch objects, handle pagination with ContinuationToken
+            list_kwargs = {'Bucket': bucket}
+            if continuation_token:
+                list_kwargs['ContinuationToken'] = continuation_token
+
```

Comment:
Nit: The error handling when fetching objects from S3 lacks specificity. You might want to raise a more detailed custom exception, so it’s easier to debug issues specific to S3 connections.